Feature/document entity permission endpoints (#45)
documaster/noark5-web-services: version1/06 - Access control.md

Access control
-------

# Introduction

> This document provides information about the access control behavior. It will help you get acquainted with the employed group-based permissions management that could be set up with the help of Access groups and Explicit and Servicve permissions.

Reviewer: Servicve -> Service

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/c60f9b579b0b87cc4567ac7ae029890b9877ccb1.

To put this into perspective, in the diagram above:
* John Doe is a member of both "Administrators" and "Managers" in the OIDC IDP
* Jane Doe is a member of "Managers" in the OIDC IDP
* Baby Doe is a member of "Users" in the OIDC IDP

With the above (external to Documaster) configuration in place, one can create the following three access groups in Documaster:
* Administrators
* Managers
* Users

Based on the "claims" attributes of these access groups and considering the fact that the OIDC IDP returns information about the group membership to Documaster, Documaster is able to map the Documaster-based and the IDP-based groups together and grant the described permissions to each user.

# Access Control Types

## Access Group

An AccessGroup is identified by the following attributes:

| Attribute          | Data type    | Comment                                                                                |
|--------------------|--------------|----------------------------------------------------------------------------------------|

> | id                 | integer      | Unique identifier of the access group                                                  |

Reviewer 2: Should this be long? As per comments from the other file (Web Services.md)

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/1603ef115d6ec6b472d6a3fe64956d82e892bea3.

| name               | string       | Human-readable name of the access group                                                |
| description        | string       | Description of the access group                                                        |
| claims             | string array | External user claims upon which a user will be assigned as a member of an access group |

> | globalPermissions  | string array | [Global permissions](#explicit-permissions)                                            |

Reviewer: Were explicit permissions meant to be used here? Edit: I suppose a new endpoint for assigning global permissions will be added to v1?

Author: Correct :D

| servicePermissions | string array | [Service permissions](#service-permissions)                                            |

## Explicit permissions

Explicit permissions can be read/set for certain entity types:

* Klassifikasjonssystem

> * Klass

Reviewer: I believe this should be `Klass -> Klasse`

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/61fb9dc5238d5fe850d846ca7c6f7316bd22b2a2.

* Arkiv
* Arkivdel
* Mappe (all types)
* Registrering (all types)

> These take the form of `explicitPermissions` on the entity types and `globalPermissions` on the  `AccessGroups`. The values read/set in that field determine the permissions assigned to an entity.

Reviewer 2: Minor: double whitespace in ....(the  `AccessGroups`).....

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/2ab703981f6faf916d07af8b46ef13d6fcb54779.

- `Create`
  - grants permissions to members of the specified access group to create children of the current entity
- `Update`
  - grants permissions to members of the specified access group to update fields of the children of the current entity such as `tittel`, `beskrivelse`, etc.
- `Delete`
  - grants permissions to members of the specified access group to delete children of the current entity
- `Grant`
  - grants permissions to members of the specified access group to modify the permissions assigned to children of the current entity
- `UpdateSystemManaged`
  - grants permissions to members of the specified access group to update system attributes of the children of the current entity such as `oprettet*` fields
- `Move`
  - grants permissions to members of the specified access group to move children of the current entity to other parents

Note that new explicit permissions may be added in the future.

Supported explicit permissions for Arkiv, Arkivdel, Mappe (all types), Registrering (all types) are all explicit permissions listed above except `ReadRelated`.

> Supported explicit permissions for Klassifikasjonssystem, Klass are all explicit permissions listed above.

Reviewer: I believe this should be `Klass -> Klasse`

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/61fb9dc5238d5fe850d846ca7c6f7316bd22b2a2.

> Supported explicit permissions for the `all` code list type are all explicit permissions listed above.

Reviewer 3: Is the `all` support also work in progress?

Author: Yes, it is. It will be added with the PR I will create after I merge the current one.

Supported explicit permissions for the `administrativenhet`, `mappetype`, `dokumenttype`, and `skjerming` code list types (and their list values) are only `ReadRelated`.

### Explicit permissions use cases

#### Users with global read access to the system

To grant users access to read everything in the archive, create or update their access group with the following explicit/global permissions:

- `ReadThis`
- `Read`
- `ReadRelated`

#### Global administrators of the system

> To grant users access to read and edit everything in the archive, create or update their access group with the following explicit/global permissions:

Reviewer 3: Should Move also be mentioned here? Same applies to the section below.

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/0ce6d928ef8765494267be25c17f0debfbf43d66.

- `ReadThis`
- `Read`
- `Create`
- `Update`
- `UpdateSystemManaged`
- `Grant`

You could optionally grant `Delete`, but Documaster recommends that only a small number of people (if any) have rights to delete data in the archive.

You can add `ReadRelated` in case there are any access controlling Klasser or Code lists.

#### Users with access only to a particular Fonds in the archive

To grant users access to read and edit everything only in a single Fonds in the archive, create or update their access group with the following explicit/global permissions:

- `ReadThis`

Additionally, update the Fonds you would like to grant them access to like so:

```json
POST rms/api/public/noark5/v1/permission/entity

{
    "groupId": long,
    "objectType": "Arkiv",
    "objectId": string,
    "explicitPermissions": ["ReadThis", "Read", "Create", "Update", "UpdateSystemManaged"]
}
```

You could optionally grant `Delete` on the Fonds, but Documaster recommends that only a small number of people (if any) have rights to delete data in the archive.

> Additionally, you will need to repeat the same request for all KlassifikasjonsSystemer to which the user needs to have access:

Reviewer: I think this should become `KlassifikasjonsSystemer -> Klassifikasjonssystemer`

Author: Fixed with https://github.com/documaster/noark5-web-services/pull/45/commits/c062176435684195d3a3bff16eab7fc599a92a1d.

```json
POST rms/api/public/noark5/v1/permission/entity

{
    "groupId": integer,
    "objectType": "Klassifikasjonssystem",
    "objectId": string,
    "explicitPermissions": ["ReadThis", "Read"]
}
```

> You could optionally grant `Create`, `Update`, or `Delete` for the Klassifikasjonssystemer if you would like users to be able to create, edit, or delete existing Klasser.

Reviewer 2: Typo: Klassifikasjonssystemer -> Klassifikasjonssystem?

Author: It's actually on purpose.

I wanted to use plural, but Klassifikasjonssystems sounded weird when the word is written in Norwegian.

Klassifikasjonssystemer means Classification systems (plural) in Norwegian.

Reviewer 2: Yes, I got that I just thought here we were talking for one KS, since it is the Create request which creates a single KS :) But I got it.

Finally, you will need to repeat the same request for all Code lists to which the user needs to have access:

```json
POST rms/api/public/noark5/v1/permission/code-list

{
    "groupId": long,
    "listType": "all",
    "explicitPermissions": ["ReadThis", "Read"]
}
```

## Service permissions

The service permissions attribute is found only on `AccessGroup`. The values of that attribute determine the types of actions members of an access group can perform in Documaster.

The following `servicePermissions` can be specified for `AccessGroup`:

- `Edit list values`
  - grants permissions to members of the specified access group to edit code list values
- `Journal`
  - grants permissions to members of the specified access group to journal Journalposter
- `Edit journaled`

>   - grants permissions to members of the specified access group to change the status of Journalposter from J to other

Reviewer 3: ```suggestion
  - grants permissions to members of the specified access group to edit journaled Journalposter
```
